### PR TITLE
Fix segfault in points to path alg

### LINKED ISSUE #54686: Point to Path tool causes QGIS to crash
### What is the bug or the crash?

Running the **Points to Path** tool (like shown below) makes QGIS crash (no error message, QGIS simply shuts down)

image

### Steps to reproduce the issue

I was trying to run the tool using this [dataset](https://drive.google.com/file/d/19m1Xn7J6lB757DrbLEaYUXu26_woZNv_/view?usp=sharing).

### Versions

QGIS version
3.28.10-Firenze
QGIS code revision
e2edb9c468
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.7.1
PROJ version
9.2.1
EPSG Registry database version
v10.088 (2023-05-13)
GEOS version
3.12.0-CAPI-1.18.0
SQLite version
3.41.1
PDAL version
2.5.5
PostgreSQL client version
unknown
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
cartography_tools
1.2.1
postgis_toolbox
1.0
qgis_resource_sharing
1.0.0
quick_map_services
0.19.33
SemiAutomaticClassificationPlugin
7.10.11
SRTM-Downloader
3.2.0
valuetool
3.0.17
visualist
0.8
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

Fixes https://github.com/qgis/QGIS/issues/54686

### CODE REVIEW

**Comment by nyalldawson** [comment]
What's the actual cause of the crash? A multipoint geometry with a null point member?

**Comment by agiudiceandrea** [comment]
The provided layer that causes the crash is a GPKG layer of POINT geometry type which contains both features with an associated Point geometry and features with an associated LineString geometry.

File: src/analysis/processing/qgsalgorithmpointstopaths.cpp [hunk 1]
```diff
@@ -249,14 +249,20 @@ QVariantMap QgsPointsToPathsAlgorithm::processAlgorithm( const QVariantMap &para
         const QgsMultiPoint mp( *qgsgeometry_cast< const QgsMultiPoint * >( geom ) );
         for ( auto pit = mp.const_parts_begin(); pit != mp.const_parts_end(); ++pit )
         {
-          const QgsPoint point( *qgsgeometry_cast< const QgsPoint * >( *pit ) );
-          allPoints[ groupValue ] << qMakePair( orderValue, point );
+          const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( *pit );
+          if ( point )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmpointstopaths.cpp:253]:
```suggestion
          if ( const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( *pit ) )
```

File: src/analysis/processing/qgsalgorithmpointstopaths.cpp [hunk 2]
```diff
@@ -249,14 +249,20 @@ QVariantMap QgsPointsToPathsAlgorithm::processAlgorithm( const QVariantMap &para
...
+            allPoints[ groupValue ] << qMakePair( orderValue, *point );
+          }
         }
       }
       else
       {
-        const QgsPoint point( *qgsgeometry_cast< const QgsPoint * >( geom ) );
-        allPoints[ groupValue ] << qMakePair( orderValue, point );
+        const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( geom );
+        if ( point )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmpointstopaths.cpp:262]:
```suggestion
        if ( const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( geom ) )
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Ok - looks good to me. The reality is this is only going to help one tiny part of QGIS with this invalid dataset, and there'll be undefined behaviour in LOTS of other places/tools. That dataset really needs fixing or a general workaround applied in the ogr feature iterator. :upside_down_face:

**Comment by andremano** [comment]
Thanks everyone :)

File: src/analysis/processing/qgsalgorithmpointstopaths.cpp (hunk 1 above)
Comment (pblottiere) [reply src/analysis/processing/qgsalgorithmpointstopaths.cpp]: It is indeed a more elegant solution, thanks @nyalldawson :)

**Pull request merged**